src/main/java/com/google/roomies/database/FirebaseDatabase.java

```diff
@@ -0,0 +1,158 @@
+package com.google.roomies;
+
+import static com.google.roomies.ProjectConstants.PROJECT_ID;
+import static com.google.roomies.ListingRequestParameterNames.TIMESTAMP;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.firestore.CollectionReference;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.FieldValue;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.QueryDocumentSnapshot;
+import com.google.cloud.firestore.QuerySnapshot;
+import com.google.cloud.firestore.WriteResult;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.firebase.cloud.FirestoreClient;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map; 
+
+public class FirebaseDatabase implements Database {
+  private Firestore db;
+  public FirebaseDatabase() throws IOException {
+    this.db = initializeDatabase();
+  }
+
+  /**
+  * Gets the database instance for this project. Initializes database if it
+  * has not been initialized.
+  *
+  * @return Firestore database instance
+  * @throws IOException if database initialization fails
+  */
+  @Override
+  public Firestore getDatabase() throws IOException {
+    if (this.db == null) {
+      this.db = initializeDatabase();
+    }
+    return this.db;
+  }
```

Comment:
I'm not sure this is necessary - you're effectively ensuring the singleton property in two places. If you update the access restrictions as discussed in the previous comment, then you don't have to worry about other people making another copy - the only access path is through the factory method you control. That said, doing it this way won't /break/ anything.